public/timeout.js

```diff
@@ -0,0 +1,161 @@
+var timeoutIntervalId = 0;
+var waitForDOMIntervalId = 0;
+var timeoutGUID = 0;
+var scrollTickerId = 0;
+var sessionLifetime = 1800; //in seconds, default to 30 minutes, modifiable based on access token exp when applicable
+var logoutLocation = "/"; //default logout location, modifiable by config
+var tokenInfo = {};
+
+
+/*
+ * decode Jwt token
+ */
+function parseJwt (token) {
+  var base64Url = token.split('.')[1];
+  var base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
+  var jsonPayload = decodeURIComponent(atob(base64).split('').map(function(c) {
+      return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
+  }).join(''));
+  return JSON.parse(jsonPayload);
+};
+
+/*
+ * get access token information stored in sessionStorage
+ */
+function getSessionTokenInfo() {
+  var keys = Object.keys(sessionStorage);
+  keys.forEach(function(key){
+    var obj;
+    try {
+      obj = JSON.parse(sessionStorage.getItem(key));
+    } catch(e) {
+      obj = null;
+    }
+    if (obj && obj["tokenResponse"] && obj["tokenResponse"]["access_token"]) {
+      tokenInfo = parseJwt(obj["tokenResponse"]["access_token"]);
+    }
+  });
+  console.log("token info? ", tokenInfo);
+}
+
+/*
+ * set maximum session time based on token exp value, if available
+ */
+function setSessionMaxTime() {
+  if (!tokenInfo && !tokenInfo.exp) return;
+  //JWT token exp is number of seconds (not milliseconds) since Epoch
+  var expiredDateTime = new Date(tokenInfo.exp * 1000); //javascript requires miliseconds since Epoch so need to multiple exp times 1000
+  var totalTime = (expiredDateTime.getTime() - Date.now());
+  if (totalTime <= 0) return;
+  sessionLifetime = (totalTime / 1000); //in seconds
+}
+
+function initTimeoutIdentifier() {
+   //set unique timeout countdown tracking interval id
+   timeoutGUID = _createUUID();
```

Comment:
Might make sense to use the JWT `jti` (JWT ID) attribute here, if available. the `jti` should be unique for each token

Reply:
thank you, ivan.  fixed